### ISSUE TITLE
[train] support per epoch shuffling with `prepare_dataloader` 

### ISSUE THREAD

**Issue by matthewdeng** [ray-project/ray#22108]
**Problem:** The current implementation of `train.torch.prepare_dataloader` does not support _per-epoch_ shuffling for the DDP case.

**Context from [DistributedSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler):**

> In distributed mode, calling the `set_epoch()` method at the beginning of each epoch **before** creating the `DataLoader` iterator is necessary to make shuffling work properly across multiple epochs. Otherwise, the same ordering will be always used.

**Current Status:** There is no way for the user to access the `DistributedSampler` created by `train.torch.prepare_dataloader` or to signal that the epoch has ended.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
